File: genetic-packer/src/main/java/genetic/packer/fx/calculation/impl/EdgeVolumeCalculator.java
package genetic.packer.fx.calculation.impl;


import java.util.function.Function;

import genetic.packer.fx.calculation.BoundingSizeCalculator;
import genetic.packer.fx.specification.BoundsGetter;
import javafx.geometry.Bounds;
import javaslang.Tuple2;
import javaslang.collection.Set;
import javaslang.collection.Traversable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * @author piotr.larysz
 */
@Component
public class EdgeVolumeCalculator implements Function<Traversable<Bounds>, Double> {

    @Autowired
    private Set<Tuple2<BoundsGetter, BoundsGetter>> boundsGetters;

    @Autowired
    private BoundingSizeCalculator.Creator boundingSizeCalculatorCreator;

    @Override
    public Double apply(Traversable<Bounds> allBoxesBounds) {
        return boundsGetters
            .map(boundingSizeCalculatorCreator.from(allBoxesBounds))
            .map(Double::valueOf)
            .reduce((a, b) -> a * b);
    }
}

File: genetic-packer/src/main/java/genetic/packer/fx/calculation/impl/EdgeVolumeCalculator.kt
package genetic.packer.fx.calculation.impl

import genetic.packer.fx.calculation.BoundingSizeCalculator
import genetic.packer.fx.specification.BoundsGetter
import javafx.geometry.Bounds
import org.springframework.stereotype.Component

@Component
class EdgeVolumeCalculator(
    private val boundsGetters: Set<Pair<BoundsGetter, BoundsGetter>>,
    private val boundingSizeCalculatorCreator: BoundingSizeCalculator.Creator
) : (Sequence<Bounds>) -> Double {

  override fun invoke(allBoxesBounds: Sequence<Bounds>) =
      with(boundingSizeCalculatorCreator.from(allBoxesBounds)) {
        boundsGetters
            .map { this(it) }
            .reduce { a, b -> a * b }
      }
}